embedded-graphics/src/primitives/rectangle.rs

```diff
@@ -310,8 +310,52 @@ impl<C> Drawable<C> for &Styled<Rectangle, PrimitiveStyle<C>>
 where
     C: PixelColor,
 {
-    fn draw<D: DrawTarget<C>>(self, display: &mut D) -> Result<(), D::Error> {
-        display.draw_rectangle(self)
+    fn draw<D: DrawTarget<Color = C>>(self, display: &mut D) -> Result<(), D::Error> {
+        // Noop if rectangle is zero sized or has no style
+        if self.style.is_transparent() || self.primitive.bottom_right().is_none() {
+            return Ok(());
+        }
+
+        let fill_area = self.primitive.shrink(self.style.inside_stroke_width());
+
+        // Fill rectangle
+        if let Some(fill_color) = self.style.fill_color {
+            display.fill_solid(&fill_area, fill_color)?;
+        }
+
+        // Draw stroke
+        if let Some(stroke_color) = self.style.effective_stroke_color() {
+            let stroke_width = self.style.stroke_width;
```

Comment:
Hm that's a good point. `stroke_area.size` might also become an issue. There are a couple of places where `saturating_add` would give some safety, but the `as i32` casts and some of the other maths will still cause problems. I'm not really sure what the behaviour should be here. 

We could change the drawtargets to return an error enum:

```rust
enum DisplayError<E> {
    /// Display- or hardware-specific error
    Error(E),

    /// Arithmetic overflow/underflow
    Overflow,

    /// Internal conversion between types failed
    Conversion(TryFromIntError)
}

trait DrawTarget {
    type Error;

    // ...

    fn draw_the_thing(&mut self, thing: Rectangle) -> Result<(), DisplayError<Self::Error>> {
        // ...

        Ok(())
    }
}
```

This would allow usage of `try_into` and/or the `checked_*` methods. Open to better ideas!

Reply:
I wasn't fully awake anymore while writing the comments yesterday, so I didn't explain this very well. Handling of integer overflows in general could need some more considerations, but that's not what I meant in this case.

If `stroke_with` is larger than the size of rectangle the border will be drawn in the wrong location, because the unlimited `stroke_with` is used to calculate the size and position of the border rectangles.

Introducing an error type to catch integer errors seems to be cumbersome to use and I don't think there is much value to the user to catch these errors, because there isn't really a good way to handle them. I would prefer that we try to minimize the effects these overflows can have, e.g. by using `saturating_...`, but I don't think it's possible to prevent them entirely. I wouldn't mind if these out of range values would cause graphical corruption, but they shouldn't cause panics or endless loops.

Reply:
I've refactored the impl and added some tests for these overdraw cases. It should now work properly.